feat(clipboard): Implement HTML and clear functionality

TukanDev opened the pull request:
This PR introduces ability to set and read HTML to/from clipboard, including implementation to trigger clipboard clear.
Currently the added features are locked behind `#[cfg(desktop)]` as i have no ability to test them for mobile targets.

TukanDev committed:
Implement arboard HTML features (desktop only) and ability to trigger clipboard clear

Signed-off-by: TukanDev <contact@tukandev.com>

TukanDev committed:
Update readme of clipboard plugin

Signed-off-by: TukanDev <contact@tukandev.com>

amrbashir commented on plugins/clipboard-manager/README.md:
```diff
@@ -60,9 +60,15 @@ fn main() {
+await writeHtml('<h1>Tauri is awesome!</h1>', '<h1>Tauri is awesome!</h1>')
+assert(await readHtml(), "<h1>Tauri is awesome!</h1>");
+
+await clear()
```
let's keep the example here simple

TukanDev replied:
I will revert README  example to previous one, thought it would be good to provide example showing all possible options.

amrbashir commented on plugins/clipboard-manager/guest-js/index.ts:
```diff
@@ -11,6 +11,7 @@
 import { invoke } from "@tauri-apps/api/core";
 
 type ClipResponse = Record<"plainText", { text: string }>;
+type ClipHtmlResponse = Record<"html", { html: string, alt_html: string }>;
```
The Rust enum has `rename_all = "camelCase"`
```suggestion
type ClipHtmlResponse = Record<"html", { html: string, altHtml: string }>;
```

amrbashir commented on plugins/clipboard-manager/guest-js/index.ts:
```diff
@@ -53,4 +54,60 @@ async function readText(): Promise<string> {
+      },
+    },
+  });
+}
```
we should make this optional, also use camelCase for JS APIs
```suggestion
async function writeHtml(
    html: string,
    altHtml?: string
): Promise<void> {
  return invoke("plugin:clipboard-manager|write_html", {
    data: {
      html: {
        html,
        altHtml
      },
    },
  });
}
```

amrbashir commented on plugins/clipboard-manager/guest-js/index.ts:
````diff
@@ -53,4 +54,60 @@ async function readText(): Promise<string> {
+ * ```
+ * @since 2.0.0
+ */
+async function readHtml(): Promise<object> {
````
We need a better return type here, seems like it is returning a string

TukanDev replied:
In this case i opted to returning entire object as i wanted to provide entire result back, but yea i agree as there is no way in arboard to determine if set_html() fallback is used instead of actual html alt will always be null, since arboard also does not have get_html() as it has for plaintext and images.

amrbashir replied:
hmmm, we could instead remove this function and rely on `get_text`, I think it is enough

TukanDev replied:
Will remove it right now

TukanDev replied:
Pushed a change stripping everything related to read_html from both sides. Will mark this as resolved.

amrbashir commented on plugins/clipboard-manager/guest-js/index.ts:
````diff
@@ -53,4 +54,60 @@ async function readText(): Promise<string> {
+}
+
+/**
+ * Gets the clipboard content as HTML text.
+ * @example
+ * ```typescript
+ * import { readHtml } from '@tauri-apps/plugin-clipboard-manager';
+ * const clipboardHtml = await readHtml();
+ * ```
````
example is incorrect

TukanDev replied:
Any pointers to improve it would be good as i honestly cant determine how to improve it, perhaps adding something like assert() or console.log()

amrbashir replied:
Sorry, I meant the documentation is wrong, this is the clear function and the documentation for readHTML function

TukanDev replied:
marking this as resolved, determined it can be removed.

amrbashir commented on plugins/clipboard-manager/src/desktop.rs:
```diff
@@ -44,4 +48,35 @@ impl<R: Runtime> Clipboard<R> {
+    pub fn clear(&self) -> crate::Result<()> {
+        match &self.clipboard {
+            Ok(clipboard) => {
+                clipboard.lock().unwrap().clear().unwrap();
+                Ok(())
```
We can't unwrap here, need to propagate the error
```suggestion
                clipboard.lock().unwrap().clear().map_err(Into::into)
```

amrbashir reviewed (changes requested):
Thanks for your contributions, left a few comments

amrbashir reviewed (changes requested):
Thank you, left a few comments

TukanDev committed:
Update plugins/clipboard-manager/src/desktop.rs

Propagate error for clear as requested

Co-authored-by: Amr Bashir <amr.bashir2015@gmail.com>

TukanDev committed:
Update plugins/clipboard-manager/guest-js/index.ts

Change to camelCase as requested

Co-authored-by: Amr Bashir <amr.bashir2015@gmail.com>

TukanDev committed:
Update plugins/clipboard-manager/guest-js/index.ts

use camelCase here too

Co-authored-by: Amr Bashir <amr.bashir2015@gmail.com>

TukanDev committed:
Update README.md

Revert back as mentioned in HTML PR

TukanDev committed:
Update index.ts

After discussion  readHtml() is decided to be removed. Will follow and remove corresponding rust side function too.

TukanDev committed:
Strip all other existence of read_html out as determined in HTML support PR conversation

Signed-off-by: TukanDev <contact@tukandev.com>

amrbashir commented on plugins/clipboard-manager/guest-js/index.ts:
```diff
@@ -11,6 +11,8 @@
 import { invoke } from "@tauri-apps/api/core";
 
 type ClipResponse = Record<"plainText", { text: string }>;
+// Commented this type out as it can come in handy to stay
+//type ClipHtmlResponse = Record<"html", { html: string, altHtml: string }>;
```
Let's remove, we can add back if needed

amrbashir commented on plugins/clipboard-manager/guest-js/index.ts:
```diff
@@ -53,4 +55,46 @@ async function readText(): Promise<string> {
+ * import { writeHtml, readHtml } from '@tauri-apps/plugin-clipboard-manager';
+ * await writeHtml('<h1>Tauri is awesome!</h1>', 'plaintext');
+ * await writeHtml('<h1>Tauri is awesome!</h1>', '<h1>Tauri is awesome</h1>'); // Will write "<h1>Tauri is awesome</h1>" as plain text
+ * assert(await readHtml(), '<h1>Tauri is awesome!</h1>');
```
we should use readText() here 
```suggestion
 * assert(await readText(), '<h1>Tauri is awesome!</h1>');
```

TukanDev replied:
Resolved

amrbashir commented on plugins/clipboard-manager/guest-js/index.ts:
````diff
@@ -53,4 +55,46 @@ async function readText(): Promise<string> {
+}
+
+/**
+ * Gets the clipboard content as HTML text.
+ * @example
+ * ```typescript
+ * import { readHtml } from '@tauri-apps/plugin-clipboard-manager';
+ * const clipboardHtml = await readHtml();
+ * ```
````
This documentation is still wrong

amrbashir commented on plugins/clipboard-manager/src/models.rs:
```diff
@@ -8,10 +8,12 @@ use serde::{Deserialize, Serialize};
 #[serde(rename_all = "camelCase")]
 pub enum ClipboardContents {
     PlainText { text: String },
+    Html { html: String, alt_html: Option<String> }
```
This is not needed, let's remove

TukanDev committed:
Apply requested changes v2

Signed-off-by: TukanDev <contact@tukandev.com>

amrbashir commented:
Looks like you need to run `pnpm build` and `cargo fmt` to fix formatting and also remove the extra undeeded `read_html` function

TukanDev committed:
pnpm run build and cargo fmt

Signed-off-by: TukanDev <contact@tukandev.com>

amrbashir commented on plugins/clipboard-manager/src/mobile.rs:
```diff
@@ -39,4 +39,17 @@ impl<R: Runtime> Clipboard<R> {
+        ))
+    }
+
+    pub fn read_html(&self) -> crate::Result<ClipboardContents> {
+        Err(crate::Error::Clipboard(
+            "Unsupported on this platform".to_string(),
+        ))
+    }
```
This can be removed

TukanDev replied:
Removed

amrbashir commented on plugins/clipboard-manager/src/mobile.rs:
```diff
@@ -39,4 +39,17 @@ impl<R: Runtime> Clipboard<R> {
     }
+
+    // Treat HTML as unsupported on mobile until tested
+    pub fn write_html(&self) -> crate::Result<ClipboardContents> {
```
To fix mobile CI failing
```suggestion
    pub fn write_html(&self, _kind: ClipKind) -> crate::Result<()> {
```

TukanDev committed:
Update plugins/clipboard-manager/src/mobile.rs

fix ci calling

Co-authored-by: Amr Bashir <amr.bashir2015@gmail.com>

amrbashir commented on plugins/clipboard-manager/src/mobile.rs:
```diff
@@ -39,4 +39,17 @@ impl<R: Runtime> Clipboard<R> {
     }
+
+    // Treat HTML as unsupported on mobile until tested
+    pub fn write_html(&self, _kind: ClipKind) -> crate::Result<ClipboardContents> {
```
You also need to change the return type too

```suggestion
    pub fn write_html(&self, _kind: ClipKind) -> crate::Result<()> {
```

TukanDev committed:
mobile read_html omg....

Signed-off-by: TukanDev <contact@tukandev.com>

TukanDev committed:
Merge remote-tracking branch 'origin/v2' into v2

amrbashir committed:
Update plugins/clipboard-manager/src/mobile.rs

TukanDev committed:
pnpm format....

Signed-off-by: TukanDev <contact@tukandev.com>

TukanDev committed:
Merge remote-tracking branch 'origin/v2' into v2

amrbashir committed:
error on mobile as well

amrbashir committed:
clear on mobile

amrbashir committed:
change file [skip ci]
